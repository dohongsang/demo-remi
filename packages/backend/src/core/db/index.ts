import { DataSource } from "typeorm";
import entities from "./entities";

export class Database {
  database: DataSource;

  constructor() {}

  init() {
    if (!this.database) {
      this.database = new DataSource({
        type: "postgres",
        host: process.env.POSTGRES_HOST,
        port: parseInt(process.env.POSTGRES_PORT ?? "5432"),
        username: process.env.POSTGRES_USERNAME,
        password: process.env.POSTGRES_PASSWORD,
        database: process.env.POSTGRES_DATABASE,
        entities,
        synchronize: true,
        logging: false,
      });
    }
    this.connect();
  }

  private connect() {
    if (this.database) {
      this.database
        .initialize()
        .then(() => console.log("Connected!"))
        .catch((error) => console.log("Connection Fail!", error));
    }
  }
}
