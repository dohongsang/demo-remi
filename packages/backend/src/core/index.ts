export * from "./error";
export * from "./security/password.security";
export * from "./server";
