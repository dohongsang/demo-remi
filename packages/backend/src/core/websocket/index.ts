import http from "http";
import { Service } from "typedi";
import { v4 as uuidv4 } from "uuid";
import WebSocket from "ws";

@Service()
export class Runner {
  private wsClients: any = {};

  constructor() {}

  init() {
    const server = http.createServer();
    const wsServer = new WebSocket.Server({ server });
    wsServer.on("connection", this.onConnection.bind(this));

    const port = process.env.WSS_SERVER_PORT;
    server.listen(port, () => {
      console.log(`Websocket started at ws://0.0.0.0:${port}`);
    });
  }

  onConnection(ws: WebSocket) {
    const userId = uuidv4();
    const bindFunc = (message: WebSocket.RawData) =>
      this.onMessage(userId, message);
    ws.on("message", bindFunc.bind(this));
    ws.on("error", this.onError.bind(this));
    ws.on("close", this.onClose.bind(this));
    this.wsClients[userId] = ws;
    console.log(`New Connection ${userId}`);
  }

  onMessage(userId: string, data: WebSocket.RawData) {
    this.handleMessage(data, userId);
  }

  onError() {
    console.log(`On Error Connection.`);
  }

  onClose() {
    console.log(`On Close Connection.`);
  }

  broadcastMessage(json: any, clientSentId: string) {
    const data = JSON.stringify(json);
    for (let userId in this.wsClients) {
      let client = this.wsClients[userId];
      if (client.readyState === WebSocket.OPEN && userId !== clientSentId) {
        client.send(data);
      }
    }
  }

  handleMessage(message: any, userId: string) {
    const dataFromClient = JSON.parse(message.toString());
    this.broadcastMessage(dataFromClient, userId);
  }
}
