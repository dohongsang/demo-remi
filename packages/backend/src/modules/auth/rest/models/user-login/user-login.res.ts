import { UserModel } from "../../../../user/models/user.model";

export interface UserLoginResponse extends UserModel {}
