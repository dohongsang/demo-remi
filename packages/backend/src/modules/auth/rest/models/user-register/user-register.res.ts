export interface UserRegisterResponse {
  id: string;
  accessToken: string
}
