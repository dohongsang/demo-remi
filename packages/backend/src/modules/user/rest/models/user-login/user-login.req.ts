export interface UserLoginRequest {
  email: string;
  hashPassword: string;
}
