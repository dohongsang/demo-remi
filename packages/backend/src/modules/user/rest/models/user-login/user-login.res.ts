import { User } from "../../../../models/user.model";

export interface UserLoginResponse extends User {}