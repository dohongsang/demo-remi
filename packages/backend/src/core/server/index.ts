import cors from "cors";
import dotenv from "dotenv";
import express from "express";
import http from "http";
import { Action, useExpressServer } from "routing-controllers";
import { Database } from "../db";
import { HttpErrorHandler } from "../middlewares/error.middleware";
import { Passport } from "../middlewares/passport.middleware";
import { Runner } from "../websocket";

dotenv.config();

export class Application {
  private app: express.Express;

  constructor() {}

  init(controllers: any[]) {
    this.app = express();
    this.app.use(cors());
    useExpressServer(this.app, {
      routePrefix: process.env.PREFIX_ENDPOINT,
      controllers,
      currentUserChecker: (action: Action) => action.request.user,
      middlewares: [HttpErrorHandler, cors()],
      defaultErrorHandler: false,
    });

    const passport = new Passport();
    passport.usePassportJWT();

    const websocket = new Runner();
    websocket.init();

    const database = new Database();
    database.init();
  }

  start() {
    this.app.listen(8080, function () {
      console.log("Listening on http://0.0.0.0:8080");
    });
  }
}
