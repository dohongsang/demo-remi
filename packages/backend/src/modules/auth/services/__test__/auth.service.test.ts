import { ForbiddenError, HttpError } from "routing-controllers";
import { Password } from "../../../../core";
import { UserLoginRequest } from "../../rest/models/user-login/user-login.req";
import { UserRegisterRequest } from "../../rest/models/user-register/user-register.req";
import { AuthService } from "../auth.service";

jest.mock("../../../../core/security/password.security");

const userDao = {
  findOne: jest.fn().mockResolvedValueOnce,
} as any;
const accountDao = {
  findOne: jest.fn(),
  transaction: jest.fn(),
} as any;

const username = "dohongsang@gmail.com";
const password = "1qazxsw2";

it("User register a new account, but system throw error because not insert db", async () => {
  accountDao.findOne.mockResolvedValue();
  const authService = new AuthService(userDao, accountDao);
  await expect(
    authService.userRegister({
      email: "",
      firstName: "",
      lastName: "",
      password: "",
    })
  ).rejects.toThrowError(HttpError);
});

it("User register a new account, but a account existed", async () => {
  accountDao.findOne.mockResolvedValue({
    email: username,
  });

  const authService = new AuthService(userDao, accountDao);
  await expect(
    authService.userRegister({
      email: username,
    } as any)
  ).rejects.toThrowError(ForbiddenError);
});

it("User register a new account, sucessfully", async () => {
  const passObject = new Password();
  const authService = new AuthService(userDao, accountDao);
  const encryptPassword = passObject.encryptPasswordFromClient(password);
  const userRegisterRequest: UserRegisterRequest = {
    email: username,
    firstName: "Đỗ Hồng",
    lastName: "Sang",
    password: encryptPassword,
  };

  accountDao.findOne.mockResolvedValueOnce();
  accountDao.findOne.mockResolvedValueOnce({
    id: "uuid",
    email: username,
  });
  await jest.mocked(accountDao.transaction).mockImplementation(() => {
    return;
  });

  jest
    .spyOn(passObject, "decryptPasswordFromClient")
    .mockImplementation(() => encryptPassword);
  await expect(authService.userRegister(userRegisterRequest)).toMatchSnapshot();
});

it("User login, but account is not existed", async () => {
  const passObject = new Password();
  const encryptPassword = passObject.encryptPasswordFromClient(password);
  accountDao.findOne.mockResolvedValue();

  const request: UserLoginRequest = {
    username: username,
    password: encryptPassword,
  };
  const authService = new AuthService(userDao, accountDao);
  await expect(authService.userLogin(request)).rejects.toThrowError(HttpError);
});

it("User login, but password is not match", async () => {
  const fakePass = "1qazxsw3";
  const passObject = new Password();
  const encryptPassword = passObject.encryptPasswordFromClient(password);
  const request: UserLoginRequest = {
    username: username,
    password: encryptPassword,
  };
  accountDao.findOne.mockResolvedValue({
    email: username,
  });

  jest.spyOn(passObject, "decryptPasswordFromClient").mockReturnValue(fakePass);
  jest.spyOn(passObject, "decryptPassword").mockReturnValue(password);
  const authService = new AuthService(userDao, accountDao);
  await expect(authService.userLogin(request));
});

it("User login, it successful", async () => {
  const passObject = new Password();
  const encryptPassword = passObject.encryptPasswordFromClient(password);
  const request: UserLoginRequest = {
    username: username,
    password: encryptPassword,
  };

  const hashKey = passObject.generateHashKey();
  const hashPassword = passObject.encryptPassword(password, hashKey);
  accountDao.findOne.mockResolvedValue({
    email: username,
    hash_key: hashKey,
    hash_password: hashPassword,
  });

  const authService = new AuthService(userDao, accountDao);
  await expect(authService.userLogin(request));
});
