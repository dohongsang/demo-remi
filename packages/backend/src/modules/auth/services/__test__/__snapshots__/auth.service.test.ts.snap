// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User register a new account, sucessfully 1`] = `Promise {}`;
